src/components/tl/sharp/tl_sharp_coll.c

```diff
@@ -0,0 +1,268 @@
+    size_t               count = args->dst.info.count;
+    ucc_datatype_t       dt    = args->dst.info.datatype;
+    struct sharp_coll_reduce_spec reduce_spec;
+    enum sharp_datatype sharp_type;
+    enum sharp_reduce_op op_type;
+    size_t data_size;
+    int ret;
+
+    task->super.super.status = UCC_INPROGRESS;
+
+    sharp_type = ucc_to_sharp_dtype[dt];
+    op_type    = ucc_to_sharp_reduce_op[args->reduce.predefined_op];
+    data_size  = ucc_dt_size(dt) * count;
+
+    if (!UCC_IS_INPLACE(*args)) {
+        ucc_tl_sharp_mem_register(TASK_CTX(task), args->src.info.buffer,
+                                  data_size, &task->allreduce.s_mem_h);
+    }
+    ucc_tl_sharp_mem_register(TASK_CTX(task), args->dst.info.buffer,
+                              data_size, &task->allreduce.r_mem_h);
+
+    if (!UCC_IS_INPLACE(*args)) {
+        reduce_spec.sbuf_desc.buffer.ptr        = args->src.info.buffer;
+        reduce_spec.sbuf_desc.buffer.mem_handle = task->allreduce.s_mem_h->mr;
+        reduce_spec.sbuf_desc.mem_type          = ucc_to_sharp_memtype[args->src.info.mem_type];
+    } else {
+        reduce_spec.sbuf_desc.buffer.ptr        = args->dst.info.buffer;
+        reduce_spec.sbuf_desc.buffer.mem_handle = task->allreduce.r_mem_h->mr;
+        reduce_spec.sbuf_desc.mem_type          = ucc_to_sharp_memtype[args->dst.info.mem_type];
+    }
+
+    reduce_spec.sbuf_desc.buffer.length     = data_size;
+    reduce_spec.sbuf_desc.type              = SHARP_DATA_BUFFER;
+    reduce_spec.rbuf_desc.buffer.ptr        = args->dst.info.buffer;
+    reduce_spec.rbuf_desc.buffer.length     = data_size;
+    reduce_spec.rbuf_desc.buffer.mem_handle = task->allreduce.r_mem_h->mr;
+    reduce_spec.rbuf_desc.type              = SHARP_DATA_BUFFER;
+    reduce_spec.rbuf_desc.mem_type          = ucc_to_sharp_memtype[args->dst.info.mem_type];
+    reduce_spec.aggr_mode                   = SHARP_AGGREGATION_NONE;
+    reduce_spec.length                      = count;
+    reduce_spec.dtype                       = sharp_type;
+    reduce_spec.op                          = op_type;
+
+    ret = sharp_coll_do_allreduce_nb(team->sharp_comm, &reduce_spec, &task->req_handle);
+    if (ret != SHARP_COLL_SUCCESS) {
+        tl_error(UCC_TASK_LIB(task), "sharp_coll_do_allreduce_nb failed:%s",
+                 sharp_coll_strerror(ret));
+        coll_task->super.status = UCC_ERR_NO_RESOURCE;
+        return ucc_task_complete(coll_task);
+    }
+
+    if (sharp_coll_req_test(task->req_handle) == 0) {
+        ucc_progress_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+        return UCC_OK;
+    }
+
+    sharp_coll_req_free(task->req_handle);
+    coll_task->super.status = UCC_OK;
+    return ucc_task_complete(coll_task);
+}
+
+ucc_status_t ucc_tl_sharp_allreduce_init(ucc_tl_sharp_task_t *task)
+{
+    ucc_coll_args_t *args = &TASK_ARGS(task);
+
+    if (args->mask & UCC_COLL_ARGS_FIELD_USERDEFINED_REDUCTIONS) {
+        return UCC_ERR_NOT_SUPPORTED;
```

Comment:
don't we need a check for ucc_op->sharp_op and ucc_dtype ->sharp_dt translation. If not-supported (like OP_AVG) we need to return ERR_NOT_SUPPORTED as well